Wireshark on MacOS

I installed wireshark on MacOS with
```
$ nix-env -iA nixpkgs.wireshark-cli
installing 'wireshark-cli-3.4.0'
```
Now when I'm trying to start it from UI it's not starting UI and I need to close it with force quit.
When I run it from bash with `-D` I see:
```
$ ./Wireshark -D
21:33:17.456  Capture Msg  Capture Interface List ...
21:33:17.456  Capture Dbg  sync_interface_list_open
21:33:17.460  Capture Dbg  sync_pipe_open_command
21:33:17.530  Capture Dbg  read 3 indicator: S empty value
21:33:17.531  Capture Dbg  sync_pipe_wait_for_child: wait till child closed
21:33:17.531  Capture Dbg  sync_pipe_wait_for_child: capture child closed after 0.000s
21:33:17.531  Capture Msg  Loading External Capture Interface List ...
21:33:17.532  Capture Dbg  spawn_sync: '/nix/store/z9n7zx7m6hymivhn0k4l4n3id0l5hapc-wireshark-qt-3.4.0/Applications/Wireshark.app/Contents/MacOS/extcap/androiddump' '--extcap-interfaces' '--extcap-version=3.4'
21:33:17.532  Capture Dbg  spawn_sync: '/nix/store/z9n7zx7m6hymivhn0k4l4n3id0l5hapc-wireshark-qt-3.4.0/Applications/Wireshark.app/Contents/MacOS/extcap/randpktdump' '--extcap-interfaces' '--extcap-version=3.4'
21:33:17.532  Capture Dbg  spawn_sync: '/nix/store/z9n7zx7m6hymivhn0k4l4n3id0l5hapc-wireshark-qt-3.4.0/Applications/Wireshark.app/Contents/MacOS/extcap/ciscodump' '--extcap-interfaces' '--extcap-version=3.4'
21:33:17.533  Capture Dbg  spawn_sync: '/nix/store/z9n7zx7m6hymivhn0k4l4n3id0l5hapc-wireshark-qt-3.4.0/Applications/Wireshark.app/Contents/MacOS/extcap/udpdump' '--extcap-interfaces' '--extcap-version=3.4'
21:33:17.533  Capture Dbg  spawn_sync: '/nix/store/z9n7zx7m6hymivhn0k4l4n3id0l5hapc-wireshark-qt-3.4.0/Applications/Wireshark.app/Contents/MacOS/extcap/sshdump' '--extcap-interfaces' '--extcap-version=3.4'
dyld: Library not loaded: libwiretap.11.dylib
  Referenced from: /nix/store/z9n7zx7m6hymivhn0k4l4n3id0l5hapc-wireshark-qt-3.4.0/Applications/Wireshark.app/Contents/MacOS/extcap/androiddump
  Reason: image not found
21:33:17.537  Capture Dbg  /nix/store/z9n7zx7m6hymivhn0k4l4n3id0l5hapc-wireshark-qt-3.4.0/Applications/Wireshark.app/Contents/MacOS/extcap/androiddump finished in 4.818ms
dyld: Library not loaded: libwsutil.12.dylib
  Referenced from: /nix/store/z9n7zx7m6hymivhn0k4l4n3id0l5hapc-wireshark-qt-3.4.0/Applications/Wireshark.app/Contents/MacOS/extcap/ciscodump
  Reason: image not found
21:33:17.537  Capture Dbg  spawn_sync: '/nix/store/z9n7zx7m6hymivhn0k4l4n3id0l5hapc-wireshark-qt-3.4.0/Applications/Wireshark.app/Contents/MacOS/extcap/androiddump' '--extcap-interfaces'
21:33:17.537  Capture Dbg  /nix/store/z9n7zx7m6hymivhn0k4l4n3id0l5hapc-wireshark-qt-3.4.0/Applications/Wireshark.app/Contents/MacOS/extcap/ciscodump finished in 5.022ms
dyld: Library not loaded: libwiretap.11.dylib
  Referenced from: /nix/store/z9n7zx7m6hymivhn0k4l4n3id0l5hapc-wireshark-qt-3.4.0/Applications/Wireshark.app/Contents/MacOS/extcap/randpktdump
  Reason: image not found
21:33:17.537  Capture Dbg  spawn_sync: '/nix/store/z9n7zx7m6hymivhn0k4l4n3id0l5hapc-wireshark-qt-3.4.0/Applications/Wireshark.app/Contents/MacOS/extcap/ciscodump' '--extcap-interfaces'
21:33:17.537  Capture Dbg  /nix/store/z9n7zx7m6hymivhn0k4l4n3id0l5hapc-wireshark-qt-3.4.0/Applications/Wireshark.app/Contents/MacOS/extcap/randpktdump finished in 5.508ms
dyld: Library not loaded: libwsutil.12.dylib
  Referenced from: /nix/store/z9n7zx7m6hymivhn0k4l4n3id0l5hapc-wireshark-qt-3.4.0/Applications/Wireshark.app/Contents/MacOS/extcap/udpdump
  Reason: image not found
21:33:17.538  Capture Dbg  /nix/store/z9n7zx7m6hymivhn0k4l4n3id0l5hapc-wireshark-qt-3.4.0/Applications/Wireshark.app/Contents/MacOS/extcap/udpdump finished in 4.393ms
21:33:17.538  Capture Dbg  spawn_sync: '/nix/store/z9n7zx7m6hymivhn0k4l4n3id0l5hapc-wireshark-qt-3.4.0/Applications/Wireshark.app/Contents/MacOS/extcap/udpdump' '--extcap-interfaces'
21:33:17.538  Capture Dbg  spawn_sync: '/nix/store/z9n7zx7m6hymivhn0k4l4n3id0l5hapc-wireshark-qt-3.4.0/Applications/Wireshark.app/Contents/MacOS/extcap/randpktdump' '--extcap-interfaces'
dyld: Library not loaded: libwsutil.12.dylib
  Referenced from: /nix/store/z9n7zx7m6hymivhn0k4l4n3id0l5hapc-wireshark-qt-3.4.0/Applications/Wireshark.app/Contents/MacOS/extcap/sshdump
  Reason: image not found
21:33:17.538  Capture Dbg  /nix/store/z9n7zx7m6hymivhn0k4l4n3id0l5hapc-wireshark-qt-3.4.0/Applications/Wireshark.app/Contents/MacOS/extcap/sshdump finished in 4.830ms
21:33:17.538  Capture Dbg  spawn_sync: '/nix/store/z9n7zx7m6hymivhn0k4l4n3id0l5hapc-wireshark-qt-3.4.0/Applications/Wireshark.app/Contents/MacOS/extcap/sshdump' '--extcap-interfaces'
dyld: Library not loaded: libwiretap.11.dylib
  Referenced from: /nix/store/z9n7zx7m6hymivhn0k4l4n3id0l5hapc-wireshark-qt-3.4.0/Applications/Wireshark.app/Contents/MacOS/extcap/androiddump
  Reason: image not found
21:33:17.540  Capture Dbg  /nix/store/z9n7zx7m6hymivhn0k4l4n3id0l5hapc-wireshark-qt-3.4.0/Applications/Wireshark.app/Contents/MacOS/extcap/androiddump finished in 3.607ms
dyld: Library not loaded: libwsutil.12.dylib
  Referenced from: /nix/store/z9n7zx7m6hymivhn0k4l4n3id0l5hapc-wireshark-qt-3.4.0/Applications/Wireshark.app/Contents/MacOS/extcap/ciscodump
  Reason: image not found
21:33:17.541  Capture Dbg  /nix/store/z9n7zx7m6hymivhn0k4l4n3id0l5hapc-wireshark-qt-3.4.0/Applications/Wireshark.app/Contents/MacOS/extcap/ciscodump finished in 3.686ms
dyld: Library not loaded: libwsutil.12.dylib
  Referenced from: /nix/store/z9n7zx7m6hymivhn0k4l4n3id0l5hapc-wireshark-qt-3.4.0/Applications/Wireshark.app/Contents/MacOS/extcap/udpdump
  Reason: image not found
21:33:17.541  Capture Dbg  /nix/store/z9n7zx7m6hymivhn0k4l4n3id0l5hapc-wireshark-qt-3.4.0/Applications/Wireshark.app/Contents/MacOS/extcap/udpdump finished in 3.545ms
dyld: Library not loaded: libwiretap.11.dylib
  Referenced from: /nix/store/z9n7zx7m6hymivhn0k4l4n3id0l5hapc-wireshark-qt-3.4.0/Applications/Wireshark.app/Contents/MacOS/extcap/randpktdump
  Reason: image not found
21:33:17.541  Capture Dbg  /nix/store/z9n7zx7m6hymivhn0k4l4n3id0l5hapc-wireshark-qt-3.4.0/Applications/Wireshark.app/Contents/MacOS/extcap/randpktdump finished in 3.507ms
dyld: Library not loaded: libwsutil.12.dylib
  Referenced from: /nix/store/z9n7zx7m6hymivhn0k4l4n3id0l5hapc-wireshark-qt-3.4.0/Applications/Wireshark.app/Contents/MacOS/extcap/sshdump
  Reason: image not found
21:33:17.542  Capture Dbg  /nix/store/z9n7zx7m6hymivhn0k4l4n3id0l5hapc-wireshark-qt-3.4.0/Applications/Wireshark.app/Contents/MacOS/extcap/sshdump finished in 3.731ms
21:33:17.542  Capture Dbg  extcap: completed discovery of 5 tools in 10.977ms
1. en0 (Wi-Fi)
2. awdl0
3. llw0
4. utun0
5. utun1
6. en5
7. lo0 (Loopback)
8. bridge0 (Thunderbolt Bridge)
9. en1 (Thunderbolt 1)
10. en2 (Thunderbolt 2)
11. en3 (Thunderbolt 3)
12. en4 (Thunderbolt 4)
13. gif0
14. stf0
15. ap1
```
So looks like it can't find lib directory.
```
$ ls ~/.nix-profile/lib/
dia				libboost_thread.dylib		libnixstore.dylib		libwireshark.14.dylib		libwiretap.11.dylib		libwsutil.12.dylib		wireshark
libboost_context.dylib		libnixexpr.dylib		libnixutil.dylib		libwireshark.dylib		libwiretap.dylib		libwsutil.dylib
libboost_system.dylib		libnixmain.dylib		libwireshark.14.0.0.dylib	libwiretap.11.0.0.dylib		libwsutil.12.0.0.dylib		pkgconfig
```
MacOS version is Big Sur
I'm junior nixpkgs user, so it's possible that I do something wrong. In this case can you please tell me what I need to do to make it working.

Based on my brief investigation, I noted the following issues (see diff/patch at the bottom):

1. Incorrect regex syntax in awk
2. Missing `snappy-c.h`. I had to add `snappy` for compilation to proceed
3. Missing `zstd.h`. I had to add `zstd` for compilation to proceed.

For issue number 3, I noted that when using the latest zstd version `1.4.9` I encountered a make error immediately. Based on this [macos setup shell script](https://github.com/wireshark/wireshark/blob/c4d6b0f6d718bd5ed80c46a0bb02479492c9a079/tools/macos-setup.sh#L162), I changed the zstd version to `1.4.2`.

It compiled for more than 6 hours, and I was preparing to create a pull request. Unfortuately it reached an error!!!:

```
[ 70%] Building C object epan/dissectors/CMakeFiles/dissectors.dir/packet-ncp2222.c.o
[ 70%] Building C object epan/dissectors/CMakeFiles/dissectors.dir/errno.c.o
[ 70%] Building C object epan/dissectors/CMakeFiles/dissectors.dir/packet-dcerpc-nt.c.o
[ 71%] Building C object epan/dissectors/CMakeFiles/dissectors.dir/packet-snort-config.c.o
[ 71%] Building C object epan/dissectors/CMakeFiles/dissectors.dir/usb.c.o
[ 71%] Building C object epan/dissectors/CMakeFiles/dissectors.dir/dissectors.c.o
[ 71%] Built target dissectors
make: *** [Makefile:160: all] Error 2
builder for '/nix/store/skfsg2dj3jny9v50mq3f5x5jdqg93bcp-wireshark-qt-3.4.4.drv' failed with exit code 2
error: build of '/nix/store/skfsg2dj3jny9v50mq3f5x5jdqg93bcp-wireshark-qt-3.4.4.drv' failed
```

I'm not really sure how to debug this further.

<details>
<summary>diff/patch</summary>

```patch
diff --git a/pkgs/applications/networking/sniffers/wireshark/default.nix b/pkgs/applications/networking/sniffers/wireshark/default.nix
index 249d58ceb60..2d31343e096 100644
--- a/pkgs/applications/networking/sniffers/wireshark/default.nix
+++ b/pkgs/applications/networking/sniffers/wireshark/default.nix
@@ -3,6 +3,7 @@
 , libssh, nghttp2, zlib, cmake, makeWrapper
 , withQt ? true, qt5 ? null
 , ApplicationServices, SystemConfiguration, gmp
+, snappy, zstd
 }:
 
 assert withQt  -> qt5  != null;
@@ -37,7 +38,7 @@ in stdenv.mkDerivation {
 
   buildInputs = [
     gettext pcre perl libpcap lua5 libssh nghttp2 openssl libgcrypt
-    libgpgerror gnutls geoip c-ares python3 glib zlib
+    libgpgerror gnutls geoip c-ares python3 glib zlib snappy zstd
   ] ++ optionals withQt  (with qt5; [ qtbase qtmultimedia qtsvg qttools ])
     ++ optionals stdenv.isLinux  [ libcap libnl ]
     ++ optionals stdenv.isDarwin [ SystemConfiguration ApplicationServices gmp ]
@@ -62,7 +63,7 @@ in stdenv.mkDerivation {
     mv $out/bin/Wireshark.app $out/Applications/Wireshark.app
 
     for f in $(find $out/Applications/Wireshark.app/Contents/PlugIns -name "*.so"); do
-        for dylib in $(otool -L $f | awk '/^\t*lib/ {print $1}'); do
+        for dylib in $(otool -L $f | awk '/^\t.*lib/ {print $1}'); do
             install_name_tool -change "$dylib" "$out/lib/$dylib" "$f"
         done
     done
diff --git a/pkgs/tools/compression/zstd/default.nix b/pkgs/tools/compression/zstd/default.nix
index 6394ea3338d..49f9a2adebc 100644
--- a/pkgs/tools/compression/zstd/default.nix
+++ b/pkgs/tools/compression/zstd/default.nix
@@ -7,72 +7,44 @@
 
 stdenv.mkDerivation rec {
   pname = "zstd";
-  version = "1.4.9";
+  version = "1.4.2";
 
   src = fetchFromGitHub {
-    owner = "facebook";
-    repo = "zstd";
+    sha256 = "0ihvkzp9v740sm43wan1f8ah8z29fhiwd4nxq11vp541sf66834g";
     rev = "v${version}";
-    sha256 = "18alxnym54gswsmsr5ra82q4k1q5fyzsyx0jykb2sk2nkpvx7334";
+    repo = "zstd";
+    owner = "facebook";
   };
 
-  nativeBuildInputs = [ cmake ]
-   ++ lib.optional stdenv.isDarwin fixDarwinDylibNames;
-  buildInputs = lib.optional stdenv.hostPlatform.isUnix bash;
-
-  patches = [
-    # This patches makes sure we do not attempt to use the MD5 implementation
-    # of the host platform when running the tests
-    ./playtests-darwin.patch
-  ];
-
-  postPatch = lib.optionalString (!static) ''
-    substituteInPlace build/cmake/CMakeLists.txt \
-      --replace 'message(SEND_ERROR "You need to build static library to build tests")' ""
-    substituteInPlace build/cmake/tests/CMakeLists.txt \
-      --replace 'libzstd_static' 'libzstd_shared'
-    sed -i \
-      "1aexport ${lib.optionalString stdenv.isDarwin "DY"}LD_LIBRARY_PATH=$PWD/build_/lib" \
-      tests/playTests.sh
-  '';
+  buildInputs = stdenv.lib.optional stdenv.isDarwin fixDarwinDylibNames;
 
-  cmakeFlags = [
-    "-DZSTD_BUILD_SHARED:BOOL=${if (!static) then "ON" else "OFF"}"
-    "-DZSTD_BUILD_STATIC:BOOL=${if static then "ON" else "OFF"}"
-    "-DZSTD_PROGRAMS_LINK_SHARED:BOOL=${if (!static) then "ON" else "OFF"}"
-    "-DZSTD_LEGACY_SUPPORT:BOOL=${if legacySupport then "ON" else "OFF"}"
-    "-DZSTD_BUILD_TESTS:BOOL=ON"
+  makeFlags = [
+    "ZSTD_LEGACY_SUPPORT=${if legacySupport then "1" else "0"}"
   ];
-  cmakeDir = "../build/cmake";
-  dontUseCmakeBuildDir = true;
-  preConfigure = ''
-    mkdir -p build_ && cd $_
-  '';
 
   checkInputs = [ file ];
   doCheck = true;
-  checkPhase = ''
-    runHook preCheck
-    # Patch shebangs for playTests
-    patchShebangs ../programs/zstdgrep
-    ctest -R playTests # The only relatively fast test.
-    runHook postCheck
+  preCheck = ''
+    substituteInPlace tests/playTests.sh \
+      --replace 'MD5SUM="md5 -r"' 'MD5SUM="md5sum"'
   '';
 
+  installFlags = [
+    "PREFIX=$(out)"
+  ];
+
   preInstall = ''
-    substituteInPlace ../programs/zstdgrep \
+    substituteInPlace programs/zstdgrep \
       --replace ":-grep" ":-${gnugrep}/bin/grep" \
-      --replace ":-zstdcat" ":-$bin/bin/zstdcat"
+      --replace ":-zstdcat" ":-$out/bin/zstdcat"
 
-    substituteInPlace ../programs/zstdless \
-      --replace "zstdcat" "$bin/bin/zstdcat"
+    substituteInPlace programs/zstdless \
+      --replace "zstdcat" "$out/bin/zstdcat"
   '';
 
-  outputs = [ "bin" "dev" ]
-    ++ lib.optional stdenv.hostPlatform.isUnix "man"
-    ++ [ "out" ];
+  enableParallelBuilding = true;
 
-  meta = with lib; {
+  meta = with stdenv.lib; {
     description = "Zstandard real-time compression algorithm";
     longDescription = ''
       Zstd, short for Zstandard, is a fast lossless compression algorithm,
@@ -83,11 +55,11 @@ stdenv.mkDerivation rec {
       speed is preserved and remain roughly the same at all settings, a
       property shared by most LZ compression algorithms, such as zlib.
     '';
-    homepage = "https://facebook.github.io/zstd/";
-    changelog = "https://github.com/facebook/zstd/blob/v${version}/CHANGELOG";
+    homepage = https://facebook.github.io/zstd/;
     license = with licenses; [ bsd3 ]; # Or, at your opinion, GPL-2.0-only.
 
-    platforms = platforms.all;
+    platforms = platforms.unix;
     maintainers = with maintainers; [ orivej ];
   };
 }
+
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index 016ffb025a6..24882593b91 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -23098,6 +23098,8 @@ in
   wireshark = callPackage ../applications/networking/sniffers/wireshark {
     inherit (darwin.apple_sdk.frameworks) ApplicationServices SystemConfiguration;
     libpcap = libpcap.override { withBluez = stdenv.isLinux; };
+    snappy = snappy;
+    zstd = zstd;
   };
   wireshark-qt = wireshark;
```
</details>

--- a/pkgs/applications/networking/sniffers/wireshark/default.nix
+++ b/pkgs/applications/networking/sniffers/wireshark/default.nix
@@ -9,6 +9,7 @@
 , buildPackages
 , c-ares
 , cmake
+, fixDarwinDylibNames
 , flex
 , gettext
 , glib
@@ -49,7 +50,9 @@
 , withQt ? true
 , qt6 ? null
 }:
-
+let
+  isAppBundle = withQt && stdenv.isDarwin;
+in
 assert withQt -> qt6 != null;
 
 stdenv.mkDerivation rec {
@@ -86,6 +89,8 @@ stdenv.mkDerivation rec {
   ] ++ lib.optionals withQt [
     qt6.wrapQtAppsHook
     wrapGAppsHook3
+  ] ++ lib.optionals stdenv.isDarwin [
+    fixDarwinDylibNames
   ];
 
   buildInputs = [
@@ -139,7 +144,7 @@ stdenv.mkDerivation rec {
     "-DBUILD_wireshark=${if withQt then "ON" else "OFF"}"
     # Fix `extcap` and `plugins` paths. See https://bugs.wireshark.org/bugzilla/show_bug.cgi?id=16444
     "-DCMAKE_INSTALL_LIBDIR=lib"
-    "-DENABLE_APPLICATION_BUNDLE=${if withQt && stdenv.isDarwin then "ON" else "OFF"}"
+    "-DENABLE_APPLICATION_BUNDLE=${if isAppBundle then "ON" else "OFF"}"
     "-DLEMON_C_COMPILER=cc"
   ] ++ lib.optionals (stdenv.buildPlatform != stdenv.hostPlatform) [
     "-DHAVE_C99_VSNPRINTF_EXITCODE__TRYRUN_OUTPUT="
@@ -163,19 +168,34 @@ stdenv.mkDerivation rec {
 
   postInstall = ''
     cmake --install . --prefix "''${!outputDev}" --component Development
-  '' + lib.optionalString (stdenv.isDarwin && withQt) ''
+  '' + lib.optionalString isAppBundle ''
     mkdir -p $out/Applications
     mv $out/bin/Wireshark.app $out/Applications/Wireshark.app
+  '' + lib.optionalString stdenv.isDarwin ''
+    local flags=()
+    for file in $out/lib/*.dylib; do
+      flags+=(-change @rpath/"$(basename "$file")" "$file")
+    done
 
-    for f in $(find $out/Applications/Wireshark.app/Contents/PlugIns -name "*.so"); do
-        for dylib in $(otool -L $f | awk '/^\t*lib/ {print $1}'); do
-            install_name_tool -change "$dylib" "$out/lib/$dylib" "$f"
-        done
+    for file in $out/lib/wireshark/extcap/*; do
+      if [ -L "$file" ]; then continue; fi
+      echo "$file: fixing dylib references"
+      # note that -id does nothing on binaries
+      install_name_tool -id "$file" "''${flags[@]}" "$file"
     done
   '';
 
   preFixup = ''
     qtWrapperArgs+=("''${gappsWrapperArgs[@]}")
+  '';
+
+  # This is done to remove some binary wrappers that wrapQtApps adds in *.app directories.
+  # Copying because unfortunately pointing Wireshark (when built as an appbundle) at $out/lib instead is nontrivial.
+  postFixup = lib.optionalString isAppBundle ''
+    rm -rf $out/Applications/Wireshark.app/Contents/MacOS/extcap $out/Applications/Wireshark.app/Contents/PlugIns
+    mkdir -p $out/Applications/Wireshark.app/Contents/PlugIns/wireshark
+    cp -r $out/lib/wireshark/plugins/4-2 $out/Applications/Wireshark.app/Contents/PlugIns/wireshark/4-2
+    cp -r $out/lib/wireshark/extcap $out/Applications/Wireshark.app/Contents/MacOS/extcap
   '';
 
   meta = with lib; {

